### PR TITLE
Bug 1043782 - [NFC][Flame] Cannot detect correct tag type for Mifare-Cla...

### PR DESCRIPTION
**Pull request by DimiL**
...ssic tag. r=yoshi

### CODE REVIEW

File: src/nci/NfcTagManager.cpp [hunk 1]
```diff
@@ -379,8 +401,15 @@ bool NfcTagManager::doTransceive(const std::vector<uint8_t>& command,
       uint8_t* cmd = new uint8_t[size];
       std::copy(command.begin(), command.end(), cmd);
 
-      tNFA_STATUS status = NFA_SendRawFrame(cmd, size,
-                             NFA_DM_DEFAULT_PRESENCE_CHECK_START_DELAY);
+      if (isMifareTech(tag.mTechLibNfcTypes[0])) {
+#ifdef NFCC_PN547
+        status = EXTNS_MfcTransceive(cmd, size);
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: I think you removed the definition of 'status',
wouldn't this cause compilation error?

File: src/nci/NfcTagManager.cpp [hunk 2]
```diff
@@ -991,6 +1077,14 @@ bool NfcTagManager::switchRfInterface(tNFA_INTF_TYPE rfInterface)
   return rVal;
 }
 
+bool NfcTagManager::isMifareTech(int techTypes)
+{
+#ifdef NFCC_PN547
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: is this ifdef neccesary?

File: src/nci/NfcTag.cpp [hunk 3]
```diff
@@ -233,6 +236,22 @@ void NfcTag::discoverTechnologies(tNFA_ACTIVATED& activationData)
       mTechList[mNumTechList] = TECHNOLOGY_TYPE_KOVIO_BARCODE;
       break;
 
+#ifdef NFCC_PN547
```
Comment (allstarschh) [inline src/nci/NfcTag.cpp]: is this ifdef neccesary?

File: src/nci/NfcManager.cpp [hunk 4]
```diff
@@ -184,8 +191,12 @@ bool NfcManager::initialize()
     }
   }
 
-  if (sIsNfaEnabled)
+  if (sIsNfaEnabled) {
+#ifdef NFCC_PN547
+    EXTNS_Close();
+#endif
```
Comment (allstarschh) [inline src/nci/NfcManager.cpp]: I saw many #ifdef occurrences
perhaps you could add another function to wrap this.

void externalClose() {
  #ifdef NFCC_PN547
    EXTNS_Close();
 #else 
   ....
#endif  
}

File: src/nci/NfcTagManager.cpp (hunk 1 above)
Comment (DimiL) [inline src/nci/NfcTagManager.cpp]: I have added status in the beginning of this function

File: src/nci/NfcTag.cpp (hunk 3 above)
Comment (DimiL) [inline src/nci/NfcTag.cpp]: Yes, NFC_PROTOCOL_MIFARE is only defined in T2M's libnfc-nci

File: src/nci/NfcTagManager.cpp (hunk 2 above)
Comment (DimiL) [inline src/nci/NfcTagManager.cpp]: Yes, NFA_PROTOCOL_MIFARE is only defined in T2M's libnfc-nci

File: src/nci/NfcManager.cpp [hunk 5]
```diff
@@ -695,8 +712,16 @@ static void nfaConnectionCallback(UINT8 connEvent, tNFA_CONN_EVT_DATA* eventData
...
     case NFA_ACTIVATED_EVT:
       ALOGD("%s: NFA_ACTIVATED_EVT: gIsSelectingRfInterface=%d, sIsDisabling=%d", __FUNCTION__, gIsSelectingRfInterface, sIsDisabling);
-      if (sIsDisabling || !sIsNfaEnabled)
+#ifdef NFCC_PN547
+      if (EXTNS_GetConnectFlag()) {
+        NfcTag::getInstance().setActivationState();
+        NfcTagManager::doConnectStatus(true);
+        break;
+      }
+#endif
```
Comment (allstarschh) [inline src/nci/NfcManager.cpp]: Shouldn't this ifdef moved to after if (sIsDisabling || ... ) ?

File: src/nci/NfcTagManager.cpp [hunk 6]
```diff
@@ -250,6 +253,10 @@ int NfcTagManager::reconnectWithStatus()
     retCode = reSelect(NFA_INTERFACE_ISO_DEP);
   } else if (tag.mTechLibNfcTypes[0] == NFA_PROTOCOL_T2T) {
     retCode = reSelect(NFA_INTERFACE_FRAME);
+  } else if (isMifareTech(tag.mTechLibNfcTypes[0])) {
+#ifdef NFCC_PN547
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: I saw libnfc-nci has already defined NFA_INTERFACE_MIFARE, is this #ifdef needed?

File: src/nci/NfcTagManager.cpp (hunk 1 above)
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: define the variables when used.

File: src/nci/NfcTagManager.cpp [hunk 7]
```diff
@@ -644,10 +688,20 @@ bool NfcTagManager::doNdefFormat()
 {
   ALOGD("%s: enter", __FUNCTION__);
   tNFA_STATUS status = NFA_STATUS_OK;
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: ditto

File: src/nci/NfcTagManager.cpp [hunk 8]
```diff
@@ -657,6 +711,28 @@ bool NfcTagManager::doNdefFormat()
...
+      sem_wait(&sFormatSem);
+      status = sFormatOk ? NFA_STATUS_OK : NFA_STATUS_FAILED;
+    } else {
+      ALOGE("%s: error status=%u", __FUNCTION__, status);
+      sem_destroy(&sFormatSem);
+    }
+
+    if(!sFormatOk) {
+      ALOGE("%s: Format with Second Key Failed", __FUNCTION__);
+    }
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: this if clause is not neccesary, or merge with line 726.

File: src/nci/NfcTagManager.cpp [hunk 9]
```diff
@@ -734,6 +810,7 @@ bool NfcTagManager::doMakeReadonly()
 {
   bool result = false;
   tNFA_STATUS status;
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: ditto.

File: src/nci/NfcTagManager.cpp [hunk 10]
```diff
@@ -1092,11 +1188,32 @@ bool NfcTagManager::doWrite(std::vector<uint8_t>& buf)
       ALOGD("%s: try format", __FUNCTION__);
       sem_init(&sFormatSem, 0, 0);
       sFormatOk = false;
-      status = NFA_RwFormatTag();
+
+      if (isMifareTech(tag.mTechLibNfcTypes[0])) {
+#ifdef NFCC_PN547
+        uint8_t key1[6] = {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF};
+		status = EXTNS_MfcFormatTag(key1, sizeof(key1));
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: align

File: src/nci/NfcManager.cpp (hunk 5 above)
Comment (DimiL) [inline src/nci/NfcManager.cpp]: The source code provided by NXP calling this before  |if (sIsDisabling || ... )| , I think EXTNS_GetConnectFlag won't be true if NFC is disabled.
I will move this after  |if (sIsDisabling || ... )| for better logic

File: src/nci/NfcTagManager.cpp (hunk 6 above)
Comment (DimiL) [inline src/nci/NfcTagManager.cpp]: Yes, it is only defined in T2M provides libnfc-nci

File: src/nci/NfcTagManager.cpp (hunk 8 above)
Comment (DimiL) [inline src/nci/NfcTagManager.cpp]: |!sFormatOk| cannot merge to line 726, status return OK in line 721 doesn't mean format is success, the format result |sFormatOk| will be set by doFormatStatus which is called in callback function.

File: src/nci/NfcTag.cpp [hunk 11]
```diff
@@ -525,6 +556,14 @@ void NfcTag::fillNfcTagMembers4(INfcTag* pINfcTag, tNFA_ACTIVATED& activationDat
...
       }
 
+#ifdef NFCC_PN547
+      case NFC_PROTOCOL_MIFARE: {
+        ALOGD("%s: Mifare Classic; tech A", fn);
+        actBytes.clear();
+        actBytes.push_back(mTechParams[i].param.pa.sel_rsp);
+        break;
+      }
+#endif
```
Comment (allstarschh) [inline src/nci/NfcTag.cpp]: move this after ISO_DEP

File: src/nci/NfcTagManager.cpp [hunk 12]
```diff
@@ -428,6 +466,7 @@ void NfcTagManager::doRead(std::vector<uint8_t>& buf)
 {
   ALOGD("%s: enter", __FUNCTION__);
   tNFA_STATUS status = NFA_STATUS_FAILED;
+  NfcTag& tag = NfcTag::getInstance();
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: define when used.

File: src/nci/NfcTagManager.cpp [hunk 13]
```diff
@@ -1070,6 +1165,7 @@ bool NfcTagManager::doWrite(std::vector<uint8_t>& buf)
   const int maxBufferSize = 1024;
```
Comment (allstarschh) [inline src/nci/NfcTagManager.cpp]: ditto

File: src/nci/NfcTag.cpp [hunk 14]
```diff
@@ -233,6 +236,22 @@ void NfcTag::discoverTechnologies(tNFA_ACTIVATED& activationData)
       mTechList[mNumTechList] = TECHNOLOGY_TYPE_KOVIO_BARCODE;
       break;
 
+#ifdef NFCC_PN547
+    case NFC_PROTOCOL_MIFARE:
+      ALOGD("%s: Mifare Classic detected", fn);
+      EXTNS_MfcInit(activationData);
+      mTechList[mNumTechList] = TECHNOLOGY_TYPE_ISO14443_3A;
+      {
```
Comment (allstarschh) [inline src/nci/NfcTag.cpp]: is { neccesary here?

**Pull request merged**